cpp/src/gandiva/llvm_generator.cc

```diff
@@ -263,9 +375,17 @@ Status LLVMGenerator::CodeGenExprValue(DexPtr value_expr, FieldDescriptorPtr out
   // define loop_var : start with 0, +1 after each iter
   llvm::PHINode* loop_var = builder.CreatePHI(types_->i64_type(), 2, "loop_var");
 
+  llvm::Value* position_var = loop_var;
+  if (mode >= 0) {
+    position_var = builder.CreateIntCast(
+        builder.CreateLoad(builder.CreateGEP(arg_selection_vector, loop_var),
+                           "uncasted_position_var"),
+        types_->i64_type(), true, "position_var");
+  }
+
   // The visitor can add code to both the entry/loop blocks.
   Visitor visitor(this, *fn, loop_entry, arg_addrs, arg_local_bitmaps, arg_context_ptr,
-                  loop_var);
+                  position_var);
```

Comment:
From the tests, it looked like you are collapsing the output (per the selection vector). but, the loop here indicates that you aren't collapsing.

let's discuss this in a call.

Reply:
Please be careful to avoid private discussions that may be of value to others in the community

Reply:
ok @wesm 

My thoughts on this. There are two alternatives here 

1. Collapse the output in the project/filter based on the selection vector
- This may be non-trivial since the bitmap handling gets messy.

2. Run the projector/filter without collapsing the output, and do another iteration over the output to collapse it.

I think 2 will be simpler though it requires a memcpy.

Reply:
@pravindra The number of iterations in the loop is determined by the size of selection vector. Instead of passing it as a separate variable, We have passed it as the size of ArrowRecordBatch. 
In case of filter_project_test.cpp we set it in the Record batch
` arrow::RecordBatch::Make(schema, selection_vector->GetNumSlots(), {array0, array1});`

Even in case of JNI, if we have a `SelectionVector` as an input to the project operations, we set the number of rows as the size of `SelectionVector` instead of the number of rows in `RecordBatch`

```
  public void evaluate(ArrowRecordBatch recordBatch,
                     SelectionVector selectionVector, List<ValueVector> outColumns)
        throws GandivaException {
    evaluate(selectionVector.getRecordCount(), recordBatch.getBuffers(),
        recordBatch.getBuffersLayout(),
        selectionVector.getType().getNumber(),
        selectionVector.getBuffer().memoryAddress(),
        selectionVector.getBuffer().capacity(),
        outColumns);
  }
```

Reply:
I went through the code again - the patch is collapsing/packing the records in the output data buffer (per the input selection vector). but, the same is not happening for the output validity buffer.